cupy/linalg/einsum.py

```diff
@@ -340,6 +352,13 @@ def reduced_binary_einsum(arr0, sub0, arr1, sub1, sub_others):
     sub_out = sub_b + sub_l + sub_r
     assert set(sub_out) <= set_others, 'operands should be reduced: unary sum'
 
+    if len(contract_dims) == 0:
+        if len(sub_out) == len(sub_others):
+            sub_out = sub_others
```

Comment:
It looks these two lines are not necessary?
There's a precondition `set(sub_out) <= set_others`.

Reply:
It is to ensure that the final output becomes C-contiguous.
https://github.com/cupy/cupy/pull/2928#discussion_r364575413

Reply:
Oh, I misread `set_others` as `sub_others` in the `assert` statement above.
But I don't understand how it affects C-contiguity. Could you explain?
Also, please write the purpose of these lines as a comment (also on the corresponding lines in `_use_cutensor` branch).

Reply:
When `len(sub_out) == len(sub_others)`, the output of this function **reduced_binary_einsum** becomes kind of final output of **einsum**. In that case, `sub_others` represents the modes of real final output of `einsum`. However `sub_out` could be different from `sub_others`. That is because `sub_out` is determined so that you can use **matmul** as einsum backend. That means the modes of the output might be different from the mode of final output, so that there is array transpose after calling **reduce_binary_einsum** at the end of **einsum** and the final output could be non-C-contiguos.

https://github.com/cupy/cupy/blob/c73a325dd034ee9abfac2c4af11aa9107ec89042/cupy/linalg/einsum.py#L605-L613

When you don't use **matmul** in **reduced_binary_einsum**, you can use `sub_others` as `sub_out` and the final output becomes C-contiguos.

@toslunar , please correct me if I am wrong.